include/ensmallen_bits/parallel_sgd/parallel_sgd_impl.hpp

```diff
@@ -92,7 +96,7 @@ typename MatType::elem_type>::type ParallelSGD<DecayPolicyType>::Optimize(
 
   // The order in which the functions will be visited.
   // TODO: maybe use function.Shuffle() instead?
-  arma::Col<size_t> visitationOrder = arma::linspace<arma::Col<size_t>>(0,
+  UVecType visitationOrder = linspace<UVecType>(0,
```

Comment:
Since the visitation order is entirely used on the CPU (e.g. each thread gets the next element of the visitation order and then decides what to do based on that), I don't think there can be any advantage to moving it to the GPU; so we should probably stick with `arma::uvec` here.

Reply:
I thought about this one further: I actually don't think it makes sense to use `ParallelSGD` with Bandicoot or any non-CPU matrix type.  The premise of ParallelSGD is that you use multiple threads that each update parts of the (sparse) gradient, and the idea is that because each function corresponds (mostly) to different parts of the sparse gradient, that there will not be much contention.  This implementation uses OpenMP atomic locks for update---but that doesn't make much sense for a GPU update, it will be terrible.  If we wanted ParallelSGD on the GPU, likely a fully different implementation (not using OpenMP) would be needed.

So personally I think we should revert the changes to this class.

Reply:
Makes sense, will revert `ParallelSGD`.